Step 1 - 문자열 덧셈 계산기 구현

min-queue opened the pull request:
잘부탁드립니다 리뷰어님 !!
감사합니다

min-queue committed:
docs: README.md 추가 및 요구 사항 정리

min-queue committed:
feat: 문자열 덧셈 계산기 구현

kkt219a commented on README.md:
```diff
@@ -3,10 +3,21 @@
+- [x] 자동차가 조건에 따라 움직였는지 움직이지 않았는지 테스트하고자 할 때는 isBetween(), isGreaterThan(), isLessThan() 등은 사용하지 않는다.
+
+
+## STEP 1 - 문자열 덧셈 계산기
+
+### 요구 사항
+
+- [x] 쉼표(,) 또는 콜론(:)을 구분자로 가지는 문자열을 전달하는 경우 구분자를 기준으로 분리한 각 숫자의 합을 반환 
+  - (예: “” => 0, "1,2" => 3, "1,2,3" => 6, “1,2:3” => 6)
+- [x] 앞의 기본 구분자(쉼표, 콜론) 외에 커스텀 구분자를 지정할 수 있다. 커스텀 구분자는 문자열 앞부분의 “//”와 “\n” 사이에 위치하는 문자를 커스텀 구분자로 사용한다. 
+  - 예를 들어 “//;\n1;2;3”과 같이 값을 입력할 경우 커스텀 구분자는 세미콜론(;)이며, 결과 값은 6이 반환되어야 한다.
+- [x] 문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw 한다]
```
요구사항 정리 👍🏻

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\n(.*)");
+
+    public int add(String input) {
+        if (input == null || input.trim().isEmpty()) {
```
문자열의 공백 문자를 제거한 후, 비어있는지 확인하는 의도가 있을까요?
공백문자로만 이뤄졌는지 확인하는 게 목적이라면 `isBlank()` 메소드로도 괜찮지 않을까 생각되서요 ☺️

참고: https://lazypazy.tistory.com/307

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+
+public class StringCalculator {
+    private static final String DEFAULT_DELIMITERS = ",|:";
+    private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile("//(.)\n(.*)");
```
패턴 캐싱 👍🏻👍🏻

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+
+import static io.micrometer.common.util.StringUtils.isBlank;
+
+public class StringCalculator {
```
현재 문자열 계산기가 값을 더하고, 수 검증, 문자열 분리 및 변환 등 많은 책임을 가지고 있는데, 역할을 적절하게 분리해 보면 어떨까요? 😃

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+
+    private void validateNoNegativeNumbers(String[] numbers) {
+        List<Integer> negativeNumbers = Arrays.stream(numbers)
+                .map(Integer::parseInt)
```
파싱에 실패할 경우 예외가 조금 더 구체적이면 좋을 것 같아요!

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    private int calculateSum(String[] numbers) {
```
음수 검증과 합계산에서 파싱이 이중으로 발생하는데,
역할을 분리할 때 일급컬렉션과 원시값 포장을 활용해보면 좋을 것 같아요!
참고: https://jojoldu.tistory.com/412

kkt219a commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,64 @@
+    @ParameterizedTest
+    @ValueSource(strings = {"1"})
+    void oneNumber(final String text) {
+        assertThat(calculator.add(text)).isSameAs(Integer.parseInt(text));
```
검증부를 하드코딩해보면 어떨까요?😃

https://jojoldu.tistory.com/615

kkt219a reviewed (changes requested):
안녕하세요 민규님 ! 😊
1단계 미션 잘 진행해주셨어요! 🔥
몇가지 코멘트를 남겼는데 확인 부탁드려요!
궁금하신 점 있으시면 편하게 댓글이나 DM으로 연락주세요.☺️

kimmingyu-runable committed:
refact: 요구 사항 반영

kkt219a commented on src/main/java/calculator/DelimiterParser.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class DelimiterParser {
```
구분자 파싱에 대한 역할 분리 👍🏻👍🏻

kkt219a commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,33 @@
+package calculator;
+
+public class Number {
```
양수에 대한 역할 분리 👍🏻👍🏻

kkt219a commented on src/main/java/calculator/NumberGroups.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class NumberGroups {
```
일급컬렉션을 활용해보셨군요 👍🏻👍🏻

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,28 @@
+
+public class StringCalculator {
+    public int add(String input) {
+        DelimiterParser delimiterParser = new DelimiterParser();
```
메소드 실행때마다 DelimeterParser를 계속 생성되고 있어요!
DelimeterParser를 재활용할 수 있는 방법들은 어떤 게 있을까요?😊

kkt219a commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,28 @@
+    public int add(String input) {
+        DelimiterParser delimiterParser = new DelimiterParser();
+
+        if (isEmptyString(input)) {
+            return 0;
+        }
```
비어있는 경우만 StringCalculator가 관리하는 의도가 있을까요?
표현해주신 NumberGroups를 통해 Number가 0인 것 하나로 표현할 수 있지 않을까 생각이 들어 여쭤봅니다 ㅎㅎ!

kkt219a commented on src/main/java/calculator/NumberGroups.java:
```diff
@@ -0,0 +1,20 @@
+                .toList();
+    }
+
+    public int sum() {
```
수를 Number로 표현하셨으니 NumberGroups의 합은 Number로 표현해보면 어떨까요? :)

kkt219a reviewed (changes requested):
안녕하세요 민규님 ! 😊 
피드백 잘 반영해주셨어요! 💯
간단한 코멘트를 남겼는데 확인 부탁드려요!
궁금하신 점 있으시면 편하게 댓글이나 DM으로 연락주세요.😃

kimmingyu-runable committed:
refact: 요구 사항 리팩터링 - 2
DelimiterParser 재활용 하는 형태로 변경,
StringCalculator 에서 빈 스트링 값을 체크하는 역활을 분리,
NumberGroups의 합을 Number로 리턴값 변경

kkt219a commented on src/main/java/calculator/DelimiterParser.java:
```diff
@@ -0,0 +1,39 @@
+    private static final String DEFAULT_DELIMITER = "[,:]";
+    private static final Pattern CUSTOM_PATTERN = Pattern.compile("//(.)\n(.*)");
+
+    private DelimiterParser() {}
+
+    public static DelimiterParser getInstance() {
+        return SingletonHolder.INSTANCE;
+    }
+
+    private static class SingletonHolder {
+        private static final DelimiterParser INSTANCE = new DelimiterParser();
+    }
```
`DelimeterParser`를 싱글톤을 사용해 재사용을 보장하셨군요 👍🏻👍🏻

kkt219a commented on src/main/java/calculator/NumberValidator.java:
```diff
@@ -0,0 +1,39 @@
+package calculator;
+
+public class NumberValidator {
+    private static final int MINIMUM_VALUE = 0;
```
매직넘버 사용 💯
Number에서 상수로 선언했던 ZERO도 활용해보면 좋을 것 같아요!

kkt219a commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Number {
+    private final int value;
+    private static final NumberValidator numberValidator = new NumberValidator();
```
수를 검증하는 책임을 별도로 분리해보셨군요.💯
NumberValidator도 매번 Number가 생성될 때마다 
새롭게 생성되는 게 아닌 재활용 할 수 있다면 더 좋을 것 같아요! 👍🏻

kkt219a commented on src/main/java/calculator/Number.java:
```diff
@@ -0,0 +1,28 @@
+        return value;
+    }
+
+    public static Number ZERO = new Number(0);
```
다른 필드들이 선언된 곳에 함께 선언하면 좋을 것 같아요!
그리고 만약 외부에서 다음 코드를 호출하면 어떻게 될까요?
```
Number.ZERO = Number.create("100");
```

kkt219a commented on src/main/java/calculator/NumberValidator.java:
```diff
@@ -0,0 +1,39 @@
+    private static final int MINIMUM_VALUE = 0;
+    public void validate(String input) {
+        if (!isBlank(input)) {
+            validateNumber(input);
+            validateNonNegative(parseNumber(input));
```
`validateNumber`와 `parseNumber`가 반환 여부만 다른데,
동일한 행위를 묶어봐도 좋을 것 같아요!

kkt219a reviewed (approved):
안녕하세요 민규님 ! 😊 
피드백 잘 반영해주셨어요! 💯
간단한 코멘트를 남겼는데, 다음 단계에서 같이 보실 수 있을 것 같아
이번 단계는 여기에서 머지하겠습니다! 다음 단계도 화이팅입니다 💪🏻💪🏻
